### PR TITLE
add support for complex cases

### PR DESCRIPTION
**Pull request by jieli-matrix**
Support has been added for complex cases. However, johnny and I found NiLang.AD should extend GVar type to thunk type.
[thunk type](https://github.com/JuliaDiff/ChainRulesCore.jl/blob/8218c2cbb24492664f5d4db6e7a41c147f5472d7/src/differentials/thunks.jl#L1)
And may be I could help to add GVar extension.
In this PR, I use unthunk to get the original type to avoid CI failed.

### CODE REVIEW

File: src/chainrules.jl [hunk 1]
```diff
@@ -1,8 +1,43 @@
-function ChainRulesCore.rrule(::typeof(*), A::AbstractSparseMatrix, B::DenseInputVecOrMat)
+function ChainRulesCore.rrule(
+    ::typeof(*), A::AbstractSparseMatrix{T}, B::DenseInputVecOrMat
+) where T 
     C = A * B
     function mul_pullback(C̄)
-        _, i_A, i_B, _, _ = (~imul!)(AD.GVar(C, C̄), AD.GVar(A), AD.GVar(B), AD.GVar(1.), AD.GVar(1.))
+        i_C, i_A, i_B, _, _ = (~imul!)(AD.GVar(C, unthunk(C̄)), AD.GVar(A), AD.GVar(B), AD.GVar(1.), AD.GVar(1.))
```
Comment (johnnychen94) [inline src/chainrules.jl]: We should instead add

```julia
@eval GVar(x::$T, y::AbstractThunk) = GVar(x, unthunk(y))
```

to https://github.com/GiggleLiu/NiLang.jl/blob/f24036c478ace467ae71f630e1db2698e38fd64d/src/autodiff/vars.jl#L72-L77

and `i_C` is an unrelated change.

File: src/chainrules.jl [hunk 2]
```diff
@@ -1,8 +1,43 @@
...
         return ChainRulesCore.NoTangent(), AD.grad(i_A), AD.grad(i_B)
     end
     return C, mul_pullback
 end
```
Comment (jieli-matrix) [inline src/chainrules.jl]: If `C̄` is thrown into `GVar`, tests broken as follows
``` julia
test_rrule: * on SparseMatrixCSC{ComplexF64, Int64},Matrix{ComplexF64}: Error During Test at /Users/lijie/.julia/packages/ChainRulesTestUtils/ugS4W/src/testers.jl:191
  Got exception outside of a @test
  MethodError: no method matching (::Inv{typeof(NiSparseArrays.imul!)})(::GVar{Matrix{ComplexF64}, Thunk{ChainRulesTestUtils.var"#49#51"{Matrix{ComplexF64}}}}, ::SparseMatrixCSC{Complex{GVar{Float64, Float64}}, Int64}, ::Matrix{Complex{GVar{Float64, Float64}}}, ::GVar{Float64, Float64}, ::GVar{Float64, Float64})
  Closest candidates are:
    (::Inv{typeof(NiSparseArrays.imul!)})(::StridedVecOrMat{T} where T, ::AbstractSparseMatrix{T, Ti} where Ti, ::Union{Adjoint{var"#s832", var"#s831"} where {var"#s832", var"#s831"<:Union{LowerTriangular, UnitLowerTriangular, UnitUpperTriangular, UpperTriangular, StridedArray{T, 2} where T, BitMatrix}}, LowerTriangular, Transpose{var"#s830", var"#s829"} where {var"#s830", var"#s829"<:Union{LowerTriangular, UnitLowerTriangular, UnitUpperTriangular, UpperTriangular, StridedArray{T, 2} where T, BitMatrix}}, UnitLowerTriangular, UnitUpperTriangular, UpperTriangular, StridedMatrix{T} where T, StridedArray{T, 1} where T, BitMatrix, BitVector}, ::Number, ::Number) where T at /Users/lijie/Fudan/git/NiSparseArrays.jl/src/linalg.jl:1
    (::Inv{typeof(NiSparseArrays.imul!)})(::StridedVecOrMat{T} where T, ::Adjoint{T, var"#s4"} where var"#s4"<:(AbstractSparseMatrix{Tv, Ti} where {Tv, Ti}), ::Union{Adjoint{var"#s832", var"#s831"} where {var"#s832", var"#s831"<:Union{LowerTriangular, UnitLowerTriangular, UnitUpperTriangular, UpperTriangular, StridedArray{T, 2} where T, BitMatrix}}, LowerTriangular, Transpose{var"#s830", var"#s829"} where {var"#s830", var"#s829"<:Union{LowerTriangular, UnitLowerTriangular, UnitUpperTriangular, UpperTriangular, StridedArray{T, 2} where T, BitMatrix}}, UnitLowerTriangular, UnitUpperTriangular, UpperTriangular, StridedMatrix{T} where T, StridedArray{T, 1} where T, BitMatrix, BitVector}, ::Number, ::Number) where T at /Users/lijie/Fudan/git/NiSparseArrays.jl/src/linalg.jl:24
    (::Inv{typeof(NiSparseArrays.imul!)})(::StridedVecOrMat{T} where T, ::Union{LowerTriangular, UnitLowerTriangular, UnitUpperTriangular, UpperTriangular, StridedArray{T, 2} where T, BitMatrix}, ::Adjoint{T, var"#s6"} where var"#s6"<:(AbstractSparseMatrix{Tv, Ti} where {Tv, Ti}), ::Number, ::Number) where T at /Users/lijie/Fudan/git/NiSparseArrays.jl/src/linalg.jl:93
```

File: src/chainrules.jl (hunk 1 above)
Comment (jieli-matrix) [reply src/chainrules.jl]: Sorry... `i_C` is just used only in debug mode... I will delete it!

**Comment by GiggleLiu** [comment]
Can you explain why thunk should be supported in NiLang with using cases?

File: src/chainrules.jl [hunk 3]
```diff
@@ -1,8 +1,43 @@
...
     C = A * B
     function mul_pullback(C̄)
-        _, i_A, i_B, _, _ = (~imul!)(AD.GVar(C, C̄), AD.GVar(A), AD.GVar(B), AD.GVar(1.), AD.GVar(1.))
+        _, i_A, i_B, _, _ = (~imul!)(AD.GVar(C, unthunk(C̄)), AD.GVar(A), AD.GVar(B), AD.GVar(1.), AD.GVar(1.))
         return ChainRulesCore.NoTangent(), AD.grad(i_A), AD.grad(i_B)
     end
     return C, mul_pullback
 end
+
+function ChainRulesCore.rrule(
```
Comment (johnnychen94) [inline src/chainrules.jl:12]: Any ideas why code coverage report complains here?

**Comment by johnnychen94** [comment]
> Can you explain why thunk should be supported in NiLang with using cases?

Because we want to build on top of ChainRulesCore and make it more closely connected to other AD systems.

**Comment by GiggleLiu** [comment]
> NiLang.AD should extend GVar type to thunk type.

What does this mean? do you mean add ChainRulesCore as a dependency inof NiLang?
It does not sound like a good idea.

**Comment by johnnychen94** [comment]
It doesn't sound like a good idea to make NiLang AD an isolated AD system. ChainRulesCore is designed to be a minimal dependency so why not?

Why would we take so much effort to add these `rrule`s here and then say it is not a good idea to not making NiLang ecosystem depending on ChainRulesCore?

**Comment by GiggleLiu** [comment]
> It doesn't sound like a good idea to make NiLang AD an isolated AD system. ChainRulesCore is designed to be a minimal dependency so why not?
> 
> Why would we take so much effort to add these `rrule`s here and then say it is not a good idea to not making NiLang ecosystem depending on ChainRulesCore?

What's the gain? (or using case). You know it is a great challenge to the infracstructure to add thunk support. I wish it pays.

**Comment by johnnychen94** [comment]
I'm expecting 

```julia
Zygote.gradient(x, y) do
    f(x, y)
end
```

calling NiLang AD backend without worrying about `GVar` or any NiLang AD internals.

No? Isn't this what this project is aimed at?

**Comment by GiggleLiu** [comment]
What is the interface now? Does a user need to `unthunk` manually?

**Comment by johnnychen94** [comment]
> https://discourse.juliacn.com/t/topic/5168#2-%E5%9F%BA%E4%BA%8E%E5%8F%AF%E9%80%86%E8%AE%A1%E7%AE%97%E7%9A%84%E7%A8%80%E7%96%8F%E7%9F%A9%E9%98%B5%E6%B1%82%E5%AF%BC
> We will port the generated backward rules to ChainRules.jl as an extension, where ChainRules.jl is the most popular Julia package providing backward rules for automatic differentiation packages.

How would you provide this support without adding ChainRulesCore dependency?

**Comment by GiggleLiu** [comment]
Maybe you have discussed a lot somewhere. But I don't know anything about the context. Do you discuss in a issue?

**Comment by GiggleLiu** [comment]
Why should it be added to NiLang.AD? I think calling `unthunk` explicitly is better than do it implicitly.
It is not about user, it is about developer. The `gradient` interface is the same.

**Comment by johnnychen94** [comment]
For performance consideration, `rrule` might return a lazy `Thunk` type as gradients Ȳ, and so to support ChainRules ecosystem we need to support `Thunk` type for NiLang's AD system `GVar`. Currently,

```julia
julia> Ȳ = ChainRules.Thunk(rand(10))
Thunk(...)

julia> GVar(rand(10), Ȳ) |> typeof
GVar{Vector{Float64}, ChainRulesCore.Thunk{Vector{Float64}}}
```

And it's clear here that `GVar(x, y::Thunk)` is no longer a matrix. But `GVar` is designed to be applied at scalar level.

**Comment by johnnychen94** [comment]
You can't expect everyone to be an expert of both NiLang AD and ChainRules and happens to have a bunch of time to spend days on finding out this trivial patch.

**Comment by johnnychen94** [comment]
Enough for this. If you decide to not having ChainRulesCore in NiLang and to not spliting NiLang.AD to a separate AD package, you've made the decision.

**Comment by GiggleLiu** [comment]
> But `GVar` is designed to be applied at scalar level.

This is true. We can not expect GVar take a generic type, otherwise, the correctness can not be guranteed. Maybe you will appreciate the level of flexibility of Zygote, it caches both scalar and tensor. But this approach pays performance. It is more than two orders slower than NiLang.AD when handling scalar AD.

**Comment by GiggleLiu** [comment]
> You can't expect everyone to be an expert of both NiLang AD and ChainRules and happens to have a bunch of time to spend days on finding out this trivial patch.

This is unfair, `thunk` is not invented by NiLang.

File: src/chainrules.jl (hunk 1 above)
Comment (GiggleLiu) [reply src/chainrules.jl]: @jieli-matrix 
Yeah, you need to add `unthunk`, this is how `ChainRules` people normally do:
https://github.com/JuliaDiff/ChainRules.jl/blob/26e46081e5210f4a81f41c9ab61ea466ca8dc006/src/rulesets/Base/evalpoly.jl#L4

You should not expect `GVar` to handle thunks for you due to the unclear gain and huge amount of work in the NiLang's infrastructure.
